### PR TITLE
Heedls 424 database delete rule change

### PR DESCRIPTION
**Pull request by JBsoftwire**

### CODE REVIEW

File: DigitalLearningSolutions.Data/DataServices/UserDataService.cs [hunk 1]
```diff
@@ -351,18 +351,6 @@ public void RemoveDelegateUser(int delegateId)
...
-                    new { delegateId }
-                );
-
-                if (existingSessions.Any())
-                {
-                    throw new UserAccountInvalidStateException(
-                        $"Delegate user id {delegateId} cannot be removed as they have already started a session."
-                        );
-                }
-
```
Comment (stellake) [inline DigitalLearningSolutions.Data/DataServices/UserDataService.cs:365]: Ah - should we Dispose in the finally block?

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests.cs [hunk 2]
```diff
@@ -318,7 +319,7 @@ public void RemoveDelegateUser_cannot_remove_delegate_user_with_started_session(
                     Action action = () => userDataService.RemoveDelegateUser(id);
 
                     // Then
-                    action.Should().Throw<UserAccountInvalidStateException>();
+                    action.Should().Throw<SqlException>();
```
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests.cs:322]: Nice - does this give a reasonable error in the logs so it's clear where/why the error came from?

File: DigitalLearningSolutions.Data/DataServices/UserDataService.cs (hunk 1 above)
Comment (JBsoftwire) [reply DigitalLearningSolutions.Data/DataServices/UserDataService.cs:365]: The dispose is handled in the catch block - dispose reverts the transaction so we don't want to fire it of there's no exception.

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests.cs (hunk 2 above)
Comment (JBsoftwire) [reply DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests.cs:322]: It does, the SQL error identifies the foreign key constraint that was violated.

**Review by stellake — APPROVED** [review]
Looks great! 👍

**Pull request merged**